File: specs/typespec/employer/cost-centers.ts
import {
	type ValidationError,
	newValidationError,
	ERR_REQUIRED,
} from "../common/common";

export type CostCenterStatus = "enabled" | "disabled";

export interface CostCenter {
	id: string;
	display_name: string;
	status: CostCenterStatus;
	notes?: string;
	created_at: string;
}

export interface AddCostCenterRequest {
	id: string;
	display_name: string;
	notes?: string;
}

export interface UpdateCostCenterRequest {
	id: string;
	display_name: string;
	status: CostCenterStatus;
	notes?: string;
}

export interface ListCostCentersRequest {
	cursor?: string;
	filter_status?: CostCenterStatus;
	limit?: number;
}

export interface ListCostCentersResponse {
	items: CostCenter[];
	next_cursor: string;
}

const ERR_ID_TOO_LONG = "id must be at most 64 characters";
const ERR_ID_INVALID =
	"id must only contain lowercase letters, numbers, hyphens, and underscores, and must start with a letter or number";
const ERR_DISPLAY_NAME_TOO_LONG = "display_name must be at most 64 characters";
const ERR_NOTES_TOO_LONG = "notes must be at most 500 characters";
const ERR_STATUS_INVALID = "status must be 'enabled' or 'disabled'";

const COST_CENTER_ID_PATTERN = /^[a-z0-9][a-z0-9_-]*$/;

const VALID_STATUSES: CostCenterStatus[] = ["enabled", "disabled"];

function validateCostCenterStatus(status: string): string | null {
	if (!VALID_STATUSES.includes(status as CostCenterStatus)) {
		return ERR_STATUS_INVALID;
	}
	return null;
}

export function validateAddCostCenterRequest(
	request: AddCostCenterRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.id) {
		errs.push(newValidationError("id", ERR_REQUIRED));
	} else if (request.id.length > 64) {
		errs.push(newValidationError("id", ERR_ID_TOO_LONG));
	} else if (!COST_CENTER_ID_PATTERN.test(request.id)) {
		errs.push(newValidationError("id", ERR_ID_INVALID));
	}

	if (!request.display_name) {
		errs.push(newValidationError("display_name", ERR_REQUIRED));
	} else if (request.display_name.length > 64) {
		errs.push(newValidationError("display_name", ERR_DISPLAY_NAME_TOO_LONG));
	}

	if (request.notes && request.notes.length > 500) {
		errs.push(newValidationError("notes", ERR_NOTES_TOO_LONG));
	}

	return errs;
}

export function validateUpdateCostCenterRequest(
	request: UpdateCostCenterRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.id) {
		errs.push(newValidationError("id", ERR_REQUIRED));
	} else if (request.id.length > 64) {
		errs.push(newValidationError("id", ERR_ID_TOO_LONG));
	} else if (!COST_CENTER_ID_PATTERN.test(request.id)) {
		errs.push(newValidationError("id", ERR_ID_INVALID));
	}

	if (!request.display_name) {
		errs.push(newValidationError("display_name", ERR_REQUIRED));
	} else if (request.display_name.length > 64) {
		errs.push(newValidationError("display_name", ERR_DISPLAY_NAME_TOO_LONG));
	}

	if (!request.status) {
		errs.push(newValidationError("status", ERR_REQUIRED));
	} else {
		const statusErr = validateCostCenterStatus(request.status);
		if (statusErr) {
			errs.push(newValidationError("status", statusErr));
		}
	}

	if (request.notes && request.notes.length > 500) {
		errs.push(newValidationError("notes", ERR_NOTES_TOO_LONG));
	}

	return errs;
}

export function validateListCostCentersRequest(
	request: ListCostCentersRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (request.filter_status) {
		const statusErr = validateCostCenterStatus(request.filter_status);
		if (statusErr) {
			errs.push(newValidationError("filter_status", statusErr));
		}
	}

	return errs;
}

File: specs/typespec/employer/cost-centers.go
package employer

import (
	"fmt"
	"regexp"

	"vetchium-api-server.typespec/common"
)

const (
	costCenterIDMaxLength          = 64
	costCenterDisplayNameMaxLength = 64
	costCenterNotesMaxLength       = 500

	errCostCenterIDRequired          = "id is required"
	errCostCenterIDTooLong           = "id must be at most 64 characters"
	errCostCenterIDInvalid           = "id must only contain lowercase letters, numbers, hyphens, and underscores, and must start with a letter or number"
	errCostCenterDisplayNameRequired = "display_name is required"
	errCostCenterDisplayNameTooLong  = "display_name must be at most 64 characters"
	errCostCenterNotesTooLong        = "notes must be at most 500 characters"
	errCostCenterStatusInvalid       = "status must be 'enabled' or 'disabled'"
)

var costCenterIDPattern = regexp.MustCompile(`^[a-z0-9][a-z0-9_-]*$`)

type CostCenterStatus string

const (
	CostCenterStatusEnabled  CostCenterStatus = "enabled"
	CostCenterStatusDisabled CostCenterStatus = "disabled"
)

// CostCenter is the response type for cost center reads.
type CostCenter struct {
	ID          string           `json:"id"`
	DisplayName string           `json:"display_name"`
	Status      CostCenterStatus `json:"status"`
	Notes       *string          `json:"notes,omitempty"`
	CreatedAt   string           `json:"created_at"`
}

// AddCostCenterRequest is the request body for POST /employer/add-cost-center.
type AddCostCenterRequest struct {
	ID          string  `json:"id"`
	DisplayName string  `json:"display_name"`
	Notes       *string `json:"notes,omitempty"`
}

func (r AddCostCenterRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.ID == "" {
		errs = append(errs, common.NewValidationError("id", fmt.Errorf(errCostCenterIDRequired)))
		return errs
	}
	if len(r.ID) > costCenterIDMaxLength {
		errs = append(errs, common.NewValidationError("id", fmt.Errorf(errCostCenterIDTooLong)))
	} else if !costCenterIDPattern.MatchString(r.ID) {
		errs = append(errs, common.NewValidationError("id", fmt.Errorf(errCostCenterIDInvalid)))
	}

	if r.DisplayName == "" {
		errs = append(errs, common.NewValidationError("display_name", fmt.Errorf(errCostCenterDisplayNameRequired)))
	} else if len(r.DisplayName) > costCenterDisplayNameMaxLength {
		errs = append(errs, common.NewValidationError("display_name", fmt.Errorf(errCostCenterDisplayNameTooLong)))
	}

	if r.Notes != nil && len(*r.Notes) > costCenterNotesMaxLength {
		errs = append(errs, common.NewValidationError("notes", fmt.Errorf(errCostCenterNotesTooLong)))
	}

	return errs
}

// UpdateCostCenterRequest is the request body for POST /employer/update-cost-center.
type UpdateCostCenterRequest struct {
	ID          string           `json:"id"`
	DisplayName string           `json:"display_name"`
	Status      CostCenterStatus `json:"status"`
	Notes       *string          `json:"notes,omitempty"`
}

func (r UpdateCostCenterRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.ID == "" {
		errs = append(errs, common.NewValidationError("id", fmt.Errorf(errCostCenterIDRequired)))
		return errs
	}
	if len(r.ID) > costCenterIDMaxLength {
		errs = append(errs, common.NewValidationError("id", fmt.Errorf(errCostCenterIDTooLong)))
	} else if !costCenterIDPattern.MatchString(r.ID) {
		errs = append(errs, common.NewValidationError("id", fmt.Errorf(errCostCenterIDInvalid)))
	}

	if r.DisplayName == "" {
		errs = append(errs, common.NewValidationError("display_name", fmt.Errorf(errCostCenterDisplayNameRequired)))
	} else if len(r.DisplayName) > costCenterDisplayNameMaxLength {
		errs = append(errs, common.NewValidationError("display_name", fmt.Errorf(errCostCenterDisplayNameTooLong)))
	}

	if r.Status == "" {
		errs = append(errs, common.NewValidationError("status", fmt.Errorf(errCostCenterStatusInvalid)))
	} else if r.Status != CostCenterStatusEnabled && r.Status != CostCenterStatusDisabled {
		errs = append(errs, common.NewValidationError("status", fmt.Errorf(errCostCenterStatusInvalid)))
	}

	if r.Notes != nil && len(*r.Notes) > costCenterNotesMaxLength {
		errs = append(errs, common.NewValidationError("notes", fmt.Errorf(errCostCenterNotesTooLong)))
	}

	return errs
}

// ListCostCentersRequest is the request body for POST /employer/list-cost-centers.
type ListCostCentersRequest struct {
	Cursor       *string           `json:"cursor,omitempty"`
	FilterStatus *CostCenterStatus `json:"filter_status,omitempty"`
	Limit        *int32            `json:"limit,omitempty"`
}

func (r ListCostCentersRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.FilterStatus != nil && *r.FilterStatus != CostCenterStatusEnabled && *r.FilterStatus != CostCenterStatusDisabled {
		errs = append(errs, common.NewValidationError("filter_status", fmt.Errorf(errCostCenterStatusInvalid)))
	}

	return errs
}

// ListCostCentersResponse is the response for POST /employer/list-cost-centers.
type ListCostCentersResponse struct {
	Items      []CostCenter `json:"items"`
	NextCursor string       `json:"next_cursor"`
}
